### ISSUE TITLE
.returning(['a', 'b']) is just returning the created id when used in .insert()

### ISSUE THREAD

**Issue by rmeloCondor** [tgriesser/knex#3038]
# Environment

Knex version: 0.16.3
Database + version: MySql (RDS) + 5.6.40
OS: AWS RDS 

# Bug

1. When I am doing an .insert() statement like db.insert().into().returning([,]) and I want to get more than one column, the statement is just returning me an array with the id created.

2. No error

**Comment by elhigu** [comment]
Returning is not supported by mysql db and also a duplicate. Last issue about this was few days ago.

**Comment by ricardograca** [comment]
Maybe it's time to re-evaluate what is more useful for users: having methods that behave the same (more or less) on all dialects or ensuring that all methods only run one query. Clearly having the `.returning()` method not work as users expect it to is a regular pain point.

**Comment by elhigu** [comment]
tl;dr I'm strictly against that (except maybe in distant future where current features of current scope are robust and consistent and even then it would be better to implement as a separate module to keep core clean, slim and robust). 

There is a feature request open about BD agnostic insert (https://github.com/tgriesser/knex/issues/2682). 

> Maybe it's time to re-evaluate what is more useful for users: having methods that behave the same (more or less) on all dialects or ensuring that all methods only run one query. 

Shifting out of that single query ideology would open path for explosion of new features, when we are already battling with maintaining library with current scope.

That would be also a path where knex would get more bloated and more bloated if we start to emulate and abstract database features but running multiple queries per query builder. IMO that should be done in higher level code in top of knex.

There are other libraries already built on top of knex, which actually provides more DB agnostic approach, like objection.js with its `graphInsert`.

Just being simple query builder is huge enough task for this size of community and specially for current codebase, where it is extremely hard to test that features work correctly under all possible conditions. 

Knex already need to talk with all different SQL databases in the world and that is a huge task in its own without trying to abstract everything over the.

> Clearly having the `.returning()` method not work as users expect it to is a regular pain point.

This is question about people not reading documentation or documentation not being clear enough. Also I would not be against of throwing an error if one calls `.returning()` with dialect that doesn't support it. 

As I have mentioned earlier I rather have error throw instead of silently building unexpected query.

**Comment by ricardograca** [comment]
I agree that silently building wrong/unsupported query is very bad. I would still like that Knex abstracted away the various database differences in terms of supported features, but I can understand wanting to keep the scope focused. That said it should at least communicated better what is and isn't supported.

**Comment by elhigu** [comment]
> I would still like that Knex abstracted away the various database differences in terms of supported features

I don't see any reason why that should be done by knex and not as a separate npm package. Biggest problem with that is currently that knex still doesn't have plugin interface to allow extending it's APIs.

**Comment by ricardograca** [comment]
> I don't see any reason why that should be done by knex and not as a separate npm package.

Because it says so on the documentation:

> Knex.js is a "batteries included" SQL query builder for Postgres, MSSQL, MySQL, MariaDB, SQLite3, Oracle, and Amazon Redshift (...) It features (...) standardized responses between different query clients and dialects.

This doesn't seem to be true considering that responses are not standardized in this case. If that mission has changed, then the documentation should be updated.

**Comment by elhigu** [comment]
> Knex.js is a "batteries included" SQL query builder for Postgres, MSSQL, MySQL, MariaDB, SQLite3, Oracle, and Amazon Redshift (...) It features (...) standardized responses between different query clients and dialects.

It actually is batteries included and standardizes those responses. But indeed that has changed from early days of knex where it tried to emulate even conflicting column types, which was a bad idea in retrospective, because it didn't really work in real world use.

I will gladly change that to something more accurate, like "standardizes responses between drivers and offers offers common API to build queries" :p

We could also be faitful for original statement by dropping support for various DB features like `.returning` which doesn't sound like a good approach. It is just a plain fact that we cannot write code that emulate every feature of every database.

I also have wanted to write down also more clear design principles, how and what knex should be and clarify more its scope to contributing documentation to not have to explain it over and over again.
